### DISCUSSION TITLE
React v18 새로나온 기능

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#27] [프론트 기술 질문 공간] [answered]
React v18에서 새로 생긴 기능 중 사용해본 기능이 있으신가요? 있으시면 어떻게 사용해보셨느지 알려주세요 
( 사용해보신 분이 많지 않을 것 같아서 다같이 한 번 살펴보면 좋을 것 같아서 넣었습니다.  사용해보지 않으신 분은 알아본 기능 중 가장 기억에 남는 기능을 소개해주세요. 다 소개해주셔도 좋습니다. )

**Comment by michoball** [comment]
Suspense

이전에는 React.lazy로 lazy loading을 하는 경우에만 가능하던 기능이 확장되었습니다. 
Suspense를 쓰면 Suspense에 감싸진 child node에 data가 fetch 되기전 까지 loading UI 컴포넌트를 보여줄 수 있는 기능을 제공합니다 .

 서버로부터 fetch 된 data를  보여주는 특정 컴포넌트만 감싸서 사용하면 다른 컴포넌트는 fetch 상태와 상관없이 UI를 보여주고 해당 컴포넌트만 fetching 되는 동안 loading 컴포넌트를 보여주기 때문에 
기존에 fetching 되는 동안 모든 컴포넌트가 렌더링을 기다리거나 혹은 loading을 위해 컴포넌트 내에 따른 jsx를 리턴했던 번거로움이 사라지게 되었습니다. 

https://react.dev/blog/2022/03/29/react-v18#new-suspense-features

**Comment by WooGie911** [comment]
React v18에서 새로 생긴 기능 중 사용해본 기능은 없지만 몇가지만 알아보자면 다음과 같습니다.

- automatic batching 을 바탕으로한 효율적인 렌더링
- 새로운 API 인 startTransition -> Suspense 와 함께 덜 중요한 상태 업데이트를 늦추는 것이 가능해짐
- 새로운 Suspense SSR 구조 -> 효율적인 hydration을 바탕으로 속도 증가

그 중 자동배치 (Automatic Batching) 에 대해서만 간단하게 설명해보겠습니다.

### New Feature : Automatic Batching

> React 에서 Batching 이란?

Batching 이란 더 나은 랜더링 퍼포먼스를 위하여 여러 state 업데이트를 (setState) 하나의 업데이트로 그룹화하는 것을 의미합니다.

React 팀이 18 버전을 출시하기 이전의 React에서는 React Event Handler만이 state 업데이트를 Batching 처리했습니다. 

하지만 18 버전 이후로는 React Event Handler뿐만 아니라 promise, setTimeout, native event handler 등 다양한 로직에서도 Batching 작업이 가능하게 되었습니다.

> 18 버전 이후 Batching 되는 방법

18 버전 이후부터는 새로운 Root API인 createRoot 를 통해서, 브라우저 이벤트 뿐만 아니라 어떤 이벤트에서 발생하는 state 업데이트를 어디서 발생했는지를 신경 쓰지 않고 자동으로 업데이트들이 배칭 시킵니다. 

모든 state 업데이트(setTimeout, promise, native event handler 등에서 발생)는 React 에서 발생하는 이벤트 내부의 업데이트와 동일한 방식으로 state 업데이트들을 Batching 하여 여러 번 수행 했어야 했던 랜더링 과정을 단 한 번만 처리할 수 있게 해줬습니다. 

이는 리랜더링 횟수를 최소화하여, 애플리케이션의 성능 향상을 기대할 수 있게 되었습니다.

> Batching 을 원하지 않는 경우

만약 Batching 을 원하지 않을 경우엔 state 업데이트를 진행할 때 ReactDOM.flusySync() 를 사용하여 Batching 하지 않고 업데이트를 진행할 수 있습니다.


참고 : https://tech.osci.kr/2022/05/03/react-18v/
참고 : https://yceffort.kr/2022/04/react-18-changelog
참고 : https://doqtqu.tistory.com/348

**Comment by michoball** [comment, marked as answer]
새로나온 기능들 중 대표적인 기능은 
Automatic Batching , Transitions, Suspense 그리고 새로운 훅들이 있겠습니다.

## Automatic Batching

React 18의 createRoot를 통해, 모든 업데이트들은 어디서 왔는가와 무관하게 자동으로 배칭되게 하는 기능입니다. 

기존에 React 이벤트 핸들러 내부에서 발생하는 업데이트만 배칭을 하였던 것에서 나아가 Promise, setTimeout, native 이벤트 핸들러, 그리고 여타 모든 이벤트 내부에서 발생하는 업데이트들도 배칭이 되도록 한 것입니다. 


## Transition

긴급 업데이트와 긴급하지 않은 업데이트로 UI 상태 업데이트를 나눠서 다루게 해주는 기능입니다. 
typing, clicking, or pressing과 같이 즉시 업데이트되어야하는 것을 긴급 업데이트라하고 이런 긴급 업데이트가 발생시 긴급하지 않은 업데이트 보다 우선적으로 업데이트를 진행하게 해줍니다. 
useTransition으로 특정 이벤트의 pending 상태에 액션을 넣을 수 있습니다.


## Suspense

컴포넌트의 랜더링을 어떤 작업이 끝날 때까지 잠시 중단시키고 다른 컴포넌트를 먼저 랜더링할 수 있도록 해주는 기능입니다. 보통 비동기 통신을 하여 데이터를 가지고오는 작업을 할 때 loading 컴포넌트를 띄우는 역할을 수행합니다. 
이전에 다양한 방법으로 loading 컴포넌트를 띄웠던 것에서 React에서 표준화를 한 기능이라고 할 수 있습니다. 

Suspense 의 fallback 속성에 loading 시 보여줄 컴포넌트를 넣어서 사용하곤 합니다. 

## 새로운 Hook 

****useId**** [ 바로가기 ](https://react.dev/blog/2022/03/29/react-v18#useid)

****useTransition****[ 바로가기 ](https://react.dev/blog/2022/03/29/react-v18#usetransition)

****useDeferredValue****[ 바로가기 ](https://react.dev/blog/2022/03/29/react-v18#usedeferredvalue)

****useSyncExternalStore****[ 바로가기 ](https://react.dev/blog/2022/03/29/react-v18#usesyncexternalstore)

****useInsertionEffect****[ 바로가기 ](https://react.dev/blog/2022/03/29/react-v18#useinsertioneffect)

등이 있습니다.
